lib/itertools/iterator.go

```diff
@@ -0,0 +1,95 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package itertools
+
+import (
+	"errors"
+)
+
+// ErrCannotReduceBatchSize is returned when an attempt is made to reduce the batch size
+// below the minimum allowed size.
+var ErrCannotReduceBatchSize = errors.New("cannot reduce batch size further")
+
+// DynamicBatchSizeIterator provides an iterator over batches of items with dynamic size adjustment.
+// It supports batch size reduction for retry scenarios via the ReduceSize() method.
+type DynamicBatchSizeIterator[T any] struct {
+	items        []T
+	currentSize  int
+	currentChunk []T
+	shouldRetry  bool
+}
+
+// DynamicBatchSize creates a new iterator that yields batches of items with dynamic size adjustment.
+//
+// Example usage:
+//
+//	chunks := itertools.DynamicBatchSize(items, 1000)
+//	for chunks.Next() {
+//	    chunk := chunks.Chunk()
+//	    if err := processBatch(chunk); err != nil {
+//	        if shouldReduceSize(err) {
+//	            if err := chunks.ReduceSize(); err != nil {
+//	                return err
+//	            }
+//	        }
+//	    }
+//	}
+func DynamicBatchSize[T any](items []T, defaultChunkSize int) *DynamicBatchSizeIterator[T] {
+	if defaultChunkSize <= 0 {
+		defaultChunkSize = 1
+	}
+	return &DynamicBatchSizeIterator[T]{
+		items:       items,
+		currentSize: defaultChunkSize,
+		shouldRetry: false,
+	}
+}
+
+// Next advances the iterator to the next batch. Returns true if there is a batch available,
+// false if the iterator is exhausted. If ReduceSize() was called on the previous iteration,
+// Next() will retry the current batch with the reduced size.
+func (it *DynamicBatchSizeIterator[T]) Next() bool {
+	// If not retrying, advance to the next chunk
+	if !it.shouldRetry && it.currentChunk != nil {
+		it.items = it.items[len(it.currentChunk):]
+	}
+	it.shouldRetry = false
+
+	if len(it.items) == 0 {
+		return false
+	}
+
+	end := min(it.currentSize, len(it.items))
+	it.currentChunk = it.items[:end]
+	return true
+}
+
+// Chunk returns the current batch of items. Should only be called after Next() returns true.
+func (it *DynamicBatchSizeIterator[T]) Chunk() []T {
+	return it.currentChunk
+}
+
+// ReduceSize reduces the batch size by half and signals that the current batch should be retried.
+// Returns ErrCannotReduceBatchSize if the batch size cannot be reduced further.
+func (it *DynamicBatchSizeIterator[T]) ReduceSize() error {
+	if len(it.currentChunk) <= 1 {
+		return ErrCannotReduceBatchSize
+	}
+	it.currentSize = (it.currentSize + 1) / 2
+	it.shouldRetry = true
```

Comment:
Instead of having an additional `shouldRetry` why not delete `currentChunk`?

Reply:
The shouldRetry explicit underlay the modeled behavior  depending on `len(currentChunk)` also works but it add more abstraction but don't have strong preference. Reworked in 5f8d9fcd7452de28b3f2d903e1c99ea9ea696086